### PR TITLE
Integrate ab-rendering

### PR DESCRIPTION
**Pull request by oliverlloyd**
## What does this change?
Offers a solution for integrating `ab-rendering` employing context and a custom hook.

## Why
By using context and a hook we arguably add more complexity but we are able to abstract this code to create a very simple api for implementing an AB test making the life of a developer who wants to add an AB test easier

### Usage:

```typescript
const ABContext = React.createContext<CoreAPI | undefined | null>(undefined);

const Example = () => {
    const AB = useAB();
    if (AB === null) return <p>IhaveNoMvtCookie</p>;
    if (AB.isUserInVariant('DummyTest', 'variant')) return <p>InTheTest</p>;
    return <p>NotInTest</p>;
};

export const WithProvider = abConfig => (
    <ABProvider {...abConfig}>
        <Example />
    </ABProvider>
);
```

### CODE REVIEW

**Comment by oliverlloyd** [comment]
[A/B Tests: Integrate into DCR](https://trello.com/c/XX5XQYQW/1789-a-b-tests-integrate-into-dcr)

File: src/model/window-guardian.ts [hunk 1]
```diff
@@ -118,6 +118,9 @@ export const makeGuardianBrowserCAPI = (CAPI: CAPIType): CAPIBrowserType => {
             discussionApiClientHeader: CAPI.config.discussionApiClientHeader,
 
             dcrSentryDsn: CAPI.config.dcrSentryDsn,
+
+            // ab tests
+            switches: CAPI.config.switches,
```
Comment (liywjl) [inline src/model/window-guardian.ts:123]: I think this is the right call, `// switches` should have never been flattered into config.

[the remark this answers is not in the document]
Comment (oliverlloyd) [reply src/model/window-guardian.ts:123]: Ahh, the do. But we're exploding the object when we pass it down

```
            isSensitive: CAPI.config.isSensitive,
            videoDuration: CAPI.config.videoDuration,
            edition: CAPI.config.edition,
            section: CAPI.config.section,
            sharedAdTargeting: CAPI.config.sharedAdTargeting, // missing type definition
            adUnit: CAPI.config.adUnit,
            discussionApiUrl: CAPI.config.discussionApiUrl,
            discussionD2Uid: CAPI.config.discussionD2Uid,
            discussionApiClientHeader: CAPI.config.discussionApiClientHeader,
```

I'll refactor that back to an object

[the remark this answers is not in the document]
Comment (oliverlloyd) [reply src/model/window-guardian.ts:112]: Okay so I've removed the explicit switches we were using. By the way, my comment above contains the wrong list, it's these that I've refactored out - a lot of them weren't being used

```
            // switches
            cmpUi: CAPI.config.switches.cmpUi,
            slotBodyEnd: CAPI.config.switches.slotBodyEnd,
            ampPrebid: CAPI.config.switches.ampPrebid,
            permutive: CAPI.config.switches.permutive,
            enableSentryReporting: CAPI.config.switches.enableSentryReporting,
            enableDiscussionSwitch: CAPI.config.switches.enableDiscussionSwitch,
```

File: src/web/lib/ab.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        </ABContext.Provider>
+    );
+};
+
+/**
+ * useAB is a wrapper around React.useContext(ABContext) to provide a
+ * check to make sure there is a ABProvider parent and throw a useful
+ * message if not
+ */
+export const useAB = () => {
```
Comment (gtrufitt) [inline src/web/lib/ab.tsx:85]: I think it's worth here type useAB as the API with a more detailed description - to improve intellisense.

image

File: src/web/lib/ab.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (gtrufitt) [reply src/web/lib/ab.tsx:1]: I guess actually, if this is moved to the library then the ADR is less required!

File: src/web/lib/ab.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+const ABContext = React.createContext<CoreAPI | undefined | null>(undefined);
+
+/**
+ * ABProvider sets an instance of ABContext
+ *
+ * Each instance of AB reads the cookie and has its own config. For DCR, you probably
+ * only need one instance but wrapping the whole App has the disadvantage of blocking
+ * on cookie read. An alternative could be to wrap sections or individual components in
+ * their own ABProviders
+ */
```
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply src/web/lib/ab.tsx:42]: You're right, it's just a few ms. Here's [a page](https://codesandbox.io/s/localstorage-vs-cookies-6w39h?file=/src/test.ts) I found with a test in it. I think the thing that I'm fixated upon with cookies is just how much slower they are compared to local storage but that's not a consideration here.

File: src/web/lib/ab.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply src/web/lib/ab.tsx:1]: I had exactly the same thought. If someone else such as apps ren wanted to use AB then they'd benefit from having these utilities directly available. And we can still export non react versions for the pure javascript projects.

**Comment by oliverlloyd** [comment]
Happy to close this PR and move the functions and tests up into AB first thing

**Pull request closed without merging**